Allow actions to run on H2 and sample database

calherries opened the pull request:
Resolves https://github.com/metabase/metabase/issues/28189

This is a blocker for the 46 release, because otherwise we'd have to disable the `:actions` features from H2. We use H2 for testing actions so we'd have to come up with a workaround for that.

Depends on [Decouple checking ddl from classifying h2 stmts](https://github.com/metabase/metabase/pull/28510/files#top)

### Changes:
To allow H2 to execute actions on H2 (except test databases, where it already works):
1. I removed `ACCESS_MODE_DATA=r` from the connection strings for H2. This command sets the `readOnly` flag on the database when it is opened, so it was preventing us from using the connections to write to the database. See more context [here](https://github.com/metabase/metabase/pull/28377#issuecomment-1438610617) for why we can't have two connections with different `ACCESS_MODE_DATA` values.
2. I executed `GRANT ALL ON SCHEMA "PUBLIC" TO GUEST` on the H2 database at `resources/sample-database.db.mv.db`, so that actions could be run on tables in the Sample Database.
Previously, all privileges are granted to the GUEST user for each table when test tables are created for test databases, but this was not the case for the sample database tables. This means new and existing instances will be able to run actions on the Sample Database. This is why the binary file for the database changed. I tested the privileges [here](https://github.com/metabase/metabase/pull/28212/files#diff-945f3c15c5c4afcb35727a530bc6203d0683f1cd042e551de9b960ae2a195e9dR105).
3. Because all connections with H2 now have write privileges, we need to block writes in native queries. I added a check that parses the SQL before it gets executed and checks if it contains any statements that write to the database (that is, everything except `SELECT`, `SHOW`, `VALUES`, `TABLE`, `CALL`, and `EXPLAIN`). That's done in a similar way to how action queries currently get checked for DDL commands.
4. I also added checks for disallowing DDL commands and default admin users before H2 action queries are executed with `execute-write-query!`. These are required for security and weren't present before.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/28212)
<!-- Reviewable:end -->

calherries commented on test/metabase/sample_data_test.clj:
```diff
@@ -99,3 +101,42 @@
+              (testing "after"
+                (is (= [1]
+                       (quantity))))
+              ;; TODO: this shouldn't be necessary, since we're modifying a temp sample database.
```
I'm going to fix this TODO tomorrow.

tsmacdonald commented on src/metabase/driver/h2.clj:
```diff
@@ -121,29 +121,47 @@
+      (let [command-type (.getCommandType command)]
+        (cond
+          ;; if the command is a CommandList, then it is a multi-statement query
+          (= (type command) org.h2.command.CommandList)
```
I assume you've tested this and figured it out, but I'm a little perplexed since `CommandList` only seems to exist in v. 1 ([javadoc link](https://javadoc.io/static/com.h2database/h2/1.2.126/org/h2/command/CommandList.html)) but according to deps.edn we're on version 2.1.212 which [does not have a `CommandList`](https://javadoc.io/static/com.h2database/h2/2.1.212/org/h2/command/package-summary.html)

calherries replied:
News to me. CommandList still exists, but it was changed from being a `public class` to just a `class` (not public). https://github.com/h2database/h2database/commit/16c099939efa010cf230187b37970eecef185137#diff-650e5c48e4aa26503577f72765d0c49f76eb197b205891048a19d1fc5f4a2fe8R16

tsmacdonald replied:
Doesn't feel great to use a package-private class, but I guess our hands are tied

calherries replied:
Maybe there's a more robust way to check whether there's multiple statements from the parser, but I couldn't find any. Before the upgrade to H2 v2.1 we were able to iterate over the commands with `(map class commands)`, but that's not possible anymore.

tsmacdonald commented on src/metabase/driver/h2.clj:
```diff
@@ -121,29 +121,47 @@
+    (when-let [command (try (.prepareCommand h2-parser query)
+                            ;; if the query is invalid, errors will get caught later
+                            (catch Throwable _ nil))]
+      (let [command-type (.getCommandType command)]
```
[The javadoc](https://javadoc.io/static/com.h2database/h2/2.1.212/org/h2/command/CommandInterface.html#getCommandType--) says this returns an `int` (that you would then compare with constants they define, e.g. `CommandInterface.SELECT` is for a select statement).

calherries replied:
Yes, I'm comparing with `CommandInterface.SELECT`. The `let` and `cond` were a bit funky before though, so I've refactored them https://github.com/metabase/metabase/pull/28212/commits/516ad7c4b9e910025018d54e44bc5da66aaf55c5

calherries replied:
I deleted the whole function, see comment below

tsmacdonald commented on src/metabase/driver/h2.clj:
```diff
@@ -121,29 +121,46 @@
+        (throw (IllegalArgumentException. "Only a single statement is allowed."))
+          ;; Command types are organized with all DDL commands listed first
+          ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
+        (< (.getCommandType command) CommandInterface/ALTER_SEQUENCE)
```
`<=`?

calherries replied:
I deleted this whole function, in favour of just checking the user is non-admin. see my comment below

tsmacdonald commented:
This seems good but I'm going to defer to people with deeper DB wizardry

calherries commented:
@tsmacdonald After your review I made a few changes. Can you check again whether it still makes sense? I decided to remove the previous DDL check entirely. 

Why? It doesn't make sense given DDL commands like `CREATE TRIGGER` aren't allowed without being an admin user, and admin users shouldn't be allowed to run queries at all. The things the DDL check were trying to prevent are prevented anyway by checking the user isn't admin.

calherries commented on test/metabase/driver/h2_test.clj:
```diff
@@ -159,31 +160,36 @@
+               :engine :h2
+               :native {:query query}})))))))
+
+(deftest check-admin-commands-test
```
This isn't related to the `metabase.driver.h2` namespace, but I wanted to confirm and document that these commands don't run on H2 databases without admin rights

camsaul reviewed (changes requested):
Seems **extremely** risky to be removing `ACCESS_MODE_DATA=r` from **all** H2 connection strings. Removing the `contains-ddl?` check would introduce a massive security hole too. We added it for a reason. Talk to @escherize about why he added it or look at the PR where he added it, #25220.

I thought we talked about having separate read-write connections for Actions vs read-only connections for general query execution? I'd feel a lot better if we did things that way.

calherries commented:
You had mentioned separating read-write and read-only connections before, but I thought parsing the query to check for select statements would be sufficient to prevent writes. I would agree it's more error prone and risky though. I'll create another PR for read-write connections for actions vs read-only connections for query execution 👍 

On removing the `contains-ddl?` check, the security hole is prevented by checking for non-admin users instead, which works since the upgrade to H2 v2. I left comments [here](https://github.com/metabase/metabase/pull/28212#issuecomment-1431692127) and [here](https://github.com/metabase/metaboat/issues/165#issuecomment-1431215298).

calherries commented on src/metabase/driver/h2.clj:
```diff
@@ -102,48 +100,16 @@
 
+(defmethod driver/execute-write-query! :h2
+  [driver query]
+  (check-native-query-not-using-default-user query)
```
Adding this `check-native-query-not-using-default-user` check is sufficient to prevent DDL tomfoolery that would be otherwise checked by `check-disallow-ddl-commands`

Commit:
Remove action subtypes from inlined-models

escherize committed:
Decouple checking ddl from classifying h2 stmts

- should enable followup for easily blocking more kinds of queries
- check all statements to make sure they aren't "ddl".

escherize committed:
fix classify-query

escherize committed:
linter fixes + get-field refactor

escherize committed:
return the CommandInterface values as ints

Reach into the CommandList when needed

escherize committed:
docstring wording

escherize committed:
catch invalid queries -- they can't be classified

Commit:
Add ddl check for `execute-write-query!`

Commit:
Remove ACCESS_MODE_DATA

Commit:
Check queries are single select statements

Commit:
Add test for sample database privileges

Commit:
Fix single-select check

Commit:
Add single-select test

Commit:
Rename and add more tests for checking read only commands

Commit:
commands -> statements

Commit:
Fix check-disallow-ddl-commands

Commit:
new line

Commit:
Add more read-only statements to the tests

calherries commented:
I still want to add more tests specific to executing actions on H2, but this is the meat of it

Commit:
Update error text

Commit:
Use are

Commit:
Add integration test for executing actions with disallowed commands

escherize commented on test/metabase/sample_data_test.clj:
```diff
@@ -99,3 +101,42 @@
+          (let [rating (fn []
+                         (->> (jdbc/query conn-spec "SELECT RATING FROM PRODUCTS WHERE PRICE = 12.345;")
+                              (map :rating)))]
+            (testing "insert row"
```
I think it's worth calling `(rating)` before the insert statement, just to highlight that the insert actually changed something.

calherries replied:
Done [Add test before inserting row](https://github.com/metabase/metabase/pull/28212/commits/c3b4925429a1a962f68c7de142ee8c17f2e7efd3)

Commit:
Add test before inserting row

Commit:
Run GRANT ALL ON SCHEMA "PUBLIC" TO GUEST

Commit:
Merge branch 'master' into h2-execute-actions

Commit:
Merge branch 'master' into h2-execute-actions

Commit:
Restore classify-query

Commit:
whitespace

bshepherdson commented on test/metabase/driver/h2_test.clj:
```diff
@@ -207,3 +207,61 @@
+      "table orders"
+      "call 1 + 1"
+      ;; Note this passes the check, but will fail on execution
+      "update venues set name = 'bill'; some query that can't be parsed;"))
```
Are we parsing the SQL, or are we asking H2 to parse it but not do anything?
What I'm driving at is, is there a possibility of drift between the parser responsible for `check-read-only-statements` and H2 itself?

john-metabase replied:
We use H2 classes to parse queries, but not execute them, and then inspect the results.

bshepherdson replied:
Ah, makes sense. Then this seems sound to me.

john-metabase commented on src/metabase/driver/h2.clj:
```diff
@@ -185,12 +185,33 @@
+  (check-read-only-statements query)
   ((get-method driver/execute-reducible-query :sql-jdbc) driver query chans respond))
 
+(defmethod driver/execute-write-query! :h2
```
You mentioned implementing a read-only connection to address the `ACCESS_MODE_DATA` concerns. Is there anything that ensures that this uses a read-only connection? I looked at the `:sql-jdbc` implementation but didn't see anything.

john-metabase replied:
Obviously, not a read-only connection for the write query... excuse me 🤦 But wherever that would be implemented.

bshepherdson replied:
The OP comment of the PR notes that dual connections is impossible, see https://github.com/metabase/metabase/pull/28377#issuecomment-1438610617 .

That's an unfortunate design from H2 but not something we can change in the near term.

john-metabase replied:
I see. Then, yeah, I'm out of my depth on the connection parameter question.

john-metabase reviewed (commented):
I don't think I'm qualified to approve with the `ACCESS_MODE_DATA` concerns but wanted to check if we're planning on using a read-only connection somehow or not.
